File: src/main/java/com/badbones69/crazycrates/api/files/AbstractConfig.java
package com.badbones69.crazycrates.api.files;

import com.badbones69.crazycrates.api.files.annotations.Comment;
import com.badbones69.crazycrates.api.files.annotations.Key;
import org.apache.commons.lang3.StringEscapeUtils;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;

public class AbstractConfig {

    private YamlConfiguration config;

    public YamlConfiguration getConfig() {
        return this.config;
    }

    public void reload(Path path, Class<? extends AbstractConfig> clazz) {
        this.config = new YamlConfiguration();

        getConfig().options().copyDefaults(true);
        getConfig().options().parseComments(true);
        getConfig().options().width(9999);

        File file = path.toFile();

        try {
            getConfig().load(file);
        } catch (IOException ignore) {
        } catch (InvalidConfigurationException e) {
            e.printStackTrace();
        }

        Arrays.stream(clazz.getDeclaredFields()).forEach(field -> {
            Key key = field.getDeclaredAnnotation(Key.class);
            Comment comment = field.getDeclaredAnnotation(Comment.class);

            if (key == null) return;

            try {
                Object classObj = getClassObject();
                Object value = getValue(key.value(), field.get(classObj));
                field.set(classObj, value instanceof String str ? StringEscapeUtils.unescapeJava(str) : value);
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }

            if (comment != null) setComments(key.value(), Arrays.stream(comment.value().split("\n")).toList());
        });

        try {
            getConfig().save(file);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    protected Object getClassObject() {
        return null;
    }

    protected Object getValue(String path, Object def) {
        if (getConfig().get(path) == null) getConfig().set(path, def);

        return getConfig().get(path);
    }

    protected void setComments(String path, List<String> comments) {
        getConfig().setComments(path, comments);
    }
}

File: common/src/main/kotlin/com/badbones69/crazycrates/files/files/AbstractConfig.kt
package com.badbones69.crazycrates.api.files

import com.badbones69.crazycrates.api.files.annotations.Comment
import com.badbones69.crazycrates.api.files.annotations.Key
import java.io.File
import java.nio.file.Path

/**
 * @author BillyGalbreath
 * @maintainer Ryder Belserion
 * @see <a href="https://github.com/BillyGalbreath">...</a>
 */
open class AbstractConfig {

    private val files: HashSet<File> = hashSetOf()

    private val configurations: HashMap<File, FileConfiguration> = hashMapOf()

    fun reload(path: Path, classObject: Any) {
        // Clear just in case.
        files.clear()
        configurations.clear()

        // Get file path
        val file = path.toFile()

        // Create file.
        if (!file.exists()) file.createNewFile()

        // Add file to hashSet.
        addFile(file)

        // Put file in hashmap.
        addConfiguration(file)

        val getConfig = getConfiguration(file)

        // Set option values.
        getConfig?.options()?.copyDefaults(true)
        getConfig?.options()?.parseComments(true)

        // Load the file.
        getConfig?.load(file)

        classObject::class.java.declaredFields.forEach { method ->
            method.isAccessible = true

            val key = method.getAnnotation(Key::class.java)
            val comment = method.getAnnotation(Comment::class.java)

            val value = getValues(file, key.value, method[classObject])

            method.set(classObject, value)

            if (comment != null) getConfiguration(file)?.setComments(key.value, comment.value.split("\n").toList())
        }

        // Save the file.
        getConfig?.save(file)
    }

    private fun addFile(file: File) = files.add(file)

    private fun addConfiguration(file: File) = configurations.put(file, YamlConfiguration.loadConfiguration(file))

    private fun getConfiguration(file: File) = configurations[file]

    private fun getValues(file: File, path: String, default: Any?): Any? {
        if (getConfiguration(file)?.get(path) == null) getConfiguration(file)?.set(path, default)

        return getConfiguration(file)?.get(path)
    }
}